src/sources/legacyaudiosourceadapter.cpp

```diff
@@ -0,0 +1,65 @@
+#include "sources/legacyaudiosourceadapter.h"
+
+#include "util/logger.h"
+
+
+namespace mixxx {
+
+namespace {
+
+const Logger kLogger("LegacyAudioSourceAdapter");
+
+} // anonymous namespace
+
+LegacyAudioSourceAdapter::LegacyAudioSourceAdapter(
+        AudioSource* pOwner,
+        LegacyAudioSource* pImpl)
+    : m_pOwner(pOwner),
+      m_pImpl(pImpl) {
+}
+
+IndexRange LegacyAudioSourceAdapter::readOrSkipSampleFrames(
+        IndexRange frameIndexRange,
+        SampleBuffer::WritableSlice* pOutputBuffer) {
+    auto readableFrames =
+            m_pOwner->adjustReadableFrameIndexRangeAndOutputBuffer(
+                    frameIndexRange, pOutputBuffer);
+    if (readableFrames.empty()) {
+        return readableFrames;
+    }
+
+    const SINT seekFrameIndex = m_pImpl->seekSampleFrame(readableFrames.head());
+    if (seekFrameIndex < readableFrames.head()) {
+        const auto precedingFrames =
+                IndexRange::between(seekFrameIndex, readableFrames.head());
+        kLogger.info()
+                << "Skipping preceding frames"
+                << precedingFrames;
+        if (precedingFrames.length() != m_pImpl->readSampleFrames(precedingFrames.length(), nullptr)) {
+            kLogger.warning()
+                    << "Failed to skip preceding frames"
+                    << precedingFrames;
+            return IndexRange();
+        }
+    }
+    DEBUG_ASSERT(seekFrameIndex >= readableFrames.head());
+
+    SINT outputSampleOffset = 0;
+    if (seekFrameIndex > readableFrames.head()) {
+        const auto unreadableFrames = readableFrames.splitHead(seekFrameIndex - readableFrames.head());
```

Comment:
It was not instantly clear for me, what this line does.
Head is in one case a single position, and in an other case a new range. the auto keyword hides this fact even more.

Can we move to head = range and top and bottom = a single index? O really other better words? Than we have the top of head.
This can than become for example cutOutHead(seekFrameIndex - readableFrames.top())